Implement several certificate properties on CurlHandler

stephentoub opened the pull request:
Adds implementations to CurlHandler for:
- ClientCertificates
- ServerCertificateValidationCallback
- CheckCertificateRevocationList

I've explicitly not added these to the contract, as we're not ready for that yet on other platforms, and I've not added tests (though I've done some testing manually, such as verifying that I could run tweaked versions of the WinHttp server callback tests against this successfully).  This is about being able to easy expose them in the contract for HttpClientHandler, hopefully in the near future.

cc: @bartonjs, @ericeil, @davidsh, @SidharthNabar, @joshfree, @kapilash 
Related to https://github.com/dotnet/corefx/issues/4476, https://github.com/dotnet/corefx/issues/4120, and https://github.com/dotnet/corefx/issues/4479

stephentoub commented on src/Common/src/System/Security/Cryptography/CertificateValidation.Unix.cs:
```diff
@@ -0,0 +1,57 @@
+{
+    internal static class CertificateValidation
+    {
+        internal static SslPolicyErrors BuildChainAndVerifyProperties(X509Chain chain, X509Certificate2 remoteCertificate, bool checkCertName, string hostName)
```
This is just refactored out of the existing CertificateValidationPal.Unix.cs file.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.SslProvider.cs:
```diff
@@ -114,42 +127,54 @@ private static void AddChannelBindingToken(X509Certificate2 certificate, IntPtr
                     using (X509Chain chain = new X509Chain())
                     {
-                        chain.ChainPolicy.RevocationMode = X509RevocationMode.NoCheck;
+                        chain.ChainPolicy.RevocationMode = easy._handler.CheckCertificateRevocationList ? X509RevocationMode.Online : X509RevocationMode.NoCheck;
```
@bartonjs, is Online rather than Offline correct here?

bartonjs replied:
Yeah, Online is "downloading new CRLs is allowed", Offline is "use the cache; if it isn't cached, fail the request".

stephentoub replied:
Ok.  It wasn't clear from the docs where Online meant "downloading new CRLs is allowed" or "download new CRLs is forced", and whether Offline meant "use the cache; if it isn't cached, fail the request" or "allow using the cache rather than forcing new CRLs to be downloaded".

stephentoub commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.SslProvider.cs:
```diff
@@ -180,8 +205,30 @@ private static int VerifyCertChain(IntPtr storeCtxPtr, IntPtr arg)
+                            {
+                                SslPolicyErrors errors = CertificateValidation.BuildChainAndVerifyProperties(
+                                    chain, leafCert, checkCertName: false, hostName: null);
+                                success = errors == SslPolicyErrors.None;
```
@bartonjs, the current checked-in implementation isn't validating the host name, so I've left it like that.  Is that correct?

davidsh replied:
>the current checked-in implementation isn't validating the host name, so I've left it like that. Is that correct?

If this code is verifying the received SSL certificate from the server, then it is supposed to verify that the DNS name used for the request "matches" the name in the certificate.  And the "match" rules are complicated since it needs to verify X509 Subject field or the Subject Alternative Field as per the RFC.  If the name doesn't match, then `SslPolicyErrors.RemoteCertificateNameMismatch` is supposed to be added to the errors.

stephentoub replied:
Ok, thanks, I'm going to change it to always do the name check, not just when the callback is being used (which is the only time the SslPolicyErrors actually matters, otherwise it's an internal implementation detail).

bartonjs replied:
@stephentoub If that's something that libcurl isn't doing automatically we'll want to backfill that to rc2.  But since that functionality isn't (or at least wasn't) part of OpenSSL I'd expect it to be in curl already.

stephentoub replied:
@bartonjs, libcurl does (https://curl.haxx.se/libcurl/c/CURLOPT_SSL_VERIFYHOST.html).  So I should change it back to avoid duplicating that?

bartonjs replied:
Having our X509CheckName used instead of theirs would allow for guaranteed consistency between no-callback and callback status codes.  So if we're going to do that we should just set CURLOPT_SSL_VERIFYHOST to 0 and always do it ourselves.

Or we could hope that they're the same logic, and only run our algorithm to build the SslPolicyErrors value.

stephentoub replied:
> So if we're going to do that we should just set CURLOPT_SSL_VERIFYHOST to 0 and always do it ourselves.

Remember, this is the slow path... by the time we get here, either there was a server callback registered so we didn't try to use OpenSSL's verification, or OpenSSL's verification failed.  The common case is we use OpenSSL's verification and it succeeds.  We do not want to force to the slow path every time, as it's, well, slow, compared to OpenSSL's verification.

bartonjs replied:
I'm fine with any of these approaches:

* Let libcurl do the validation, we only do it for a server callback
* Let libcurl do the validation when there's no callback, we do it when there is one.
* Tell libcurl to not bother, we do the validation quickly in the accelerated path, and save the answer for the callback path.
    * This wouldn't have any of the previous perf problems since we're not doing a bunch of native -> managed -> native object rehydrations.

Really the difference between the first two is "codify the behavior with a test".  If curl doing the validation always means a hostname match is blocked, we need to have them not do it when a user callback is presented, since the user could be forgiving the hostname violation.

stephentoub replied:
> Let libcurl do the validation when there's no callback, we do it when there is one.... If curl doing the validation always means a hostname match is blocked, we need to have them not do it when a user callback is presented, since the user could be forgiving the hostname violation

Yup, this is what's being done in the most recent update.  See the setting of VERIFYPEER and VERIFYHOST.

stephentoub replied:
> This wouldn't have any of the previous perf problems since we're not doing a bunch of native -> managed -> native object rehydrations.

I don't understand this.  We'd be handed the native pointer and would still need to construct from it the X509Certificate2.  Or are you saying we'd add new code paths that don't construct safe handles, just call the raw interop signatures with the raw handle, etc.?

stephentoub commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.SslProvider.cs:
```diff
@@ -180,8 +205,30 @@ private static int VerifyCertChain(IntPtr storeCtxPtr, IntPtr arg)
 
                         using (X509Certificate2 leafCert = new X509Certificate2(leafCertPtr))
```
@davidsh, the WinHttp implementation isn't disposing of this X509Certificate2, and there's a WinHttp test that's then relying on that, expecting the certificate passed to the callback to be valid even after the callback returns.  That seems wrong, so I'm disposing this, but let me know if there's a good reason this shouldn't be disposed in the callback case.

davidsh replied:
@stephentoub I'm having trouble following the lines of code that you are referring to in WinHttpHandler.  Can you clarify which lines of code in WinHttpHandler have this problem?

stephentoub replied:
@davidsh, this code here:
https://github.com/dotnet/corefx/blob/master/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpRequestCallback.cs#L260-L292
It creates the X509Certificate2 but never disposes of it, leaving it for finalization.  If, however, it is disposed, then this test code fails:
https://github.com/dotnet/corefx/blob/master/src/System.Net.Http.WinHttpHandler/tests/FunctionalTests/ServerCertificateTest.cs#L122
as it's called after the callback has already returned yet expects the certificate to not have been disposed.

kapilash commented on src/Common/src/System/Net/Http/TlsCertificateExtensions.cs:
```diff
@@ -73,6 +73,16 @@ private static bool IsClientCertificate(X509Certificate2 cert)
+            chain.ChainPolicy.RevocationFlag = X509RevocationFlag.ExcludeRoot;
+            chain.ChainPolicy.ApplicationPolicy.Add(s_clientCertOidInst);
+
+            return chain.Build(certificate) ? chain : null;
```
Nit: the c# 6 "?." operator cannot be used here?

stephentoub replied:
No, it's not applicable. I know chain isn't null here.

kapilash replied:
Right, my bad. ```Build``` returns a boolean. I forgot.

stephentoub commented on src/Common/src/System/Security/Cryptography/CertificateValidation.Unix.cs:
```diff
@@ -0,0 +1,57 @@
+                    // It also does host case normalization.  The bypass logic would be something
+                    // like "all characters being within [a-z0-9.-]+"
+                    // Since it's not documented as being thread safe, create a new one each time.
+                    string matchName = new IdnMapping().GetAscii(hostName);
```
@ellismg, @bartonjs mentioned he asked you about this previously and you suggested that we not rely on IdnMapping's members being thread-safe because they're not documented as such?  That makes sense, but it really looks like the implementation is safe today... could we instead just document that it is?

stephentoub commented:
I added a commit that does the name checking in all cases.  I also noticed some other cert-related issues, mostly related to not disposing things, which I cleaned up.

stephentoub commented:
Test Innerloop Ubuntu Debug Build and Test please (https://github.com/dotnet/corefx/issues/5660)

bartonjs commented on src/Common/src/System/Net/Http/TlsCertificateExtensions.cs:
```diff
@@ -73,6 +73,24 @@ private static bool IsClientCertificate(X509Certificate2 cert)
+            var chain = new X509Chain();
+            chain.ChainPolicy.VerificationFlags = X509VerificationFlags.AllFlags;
+            chain.ChainPolicy.RevocationFlag = X509RevocationFlag.ExcludeRoot;
+            chain.ChainPolicy.ApplicationPolicy.Add(s_clientCertOidInst);
```
This ApplicationPolicy element shouldn't be set for the single-cert case; it'll cause the chain to fail to build if the cert isn't an appropriate client auth cert, but we're going to select the leaf cert anyways.

stephentoub replied:
Why was it correct before then for the other use?  Wasn't that also a single-cert case? https://github.com/dotnet/corefx/blob/master/src/Common/src/System/Net/Http/TlsCertificateExtensions.cs#L92-L111

bartonjs replied:
In Automatic mode it has to match the presented allowed CAs list and be a valid client auth cert (so the chain is part of the filter).  In manual mode where you've given a single cert there's no filtering, just doing.

Since the single manual cert is being chosen no matter what this method does, it seems better to let the chain build so that intermediate CA certs are sent.

stephentoub replied:
Ok.  I'll parameterize the helper on whether to add the application policy.

bartonjs commented on src/Common/src/System/Security/Cryptography/CertificateValidation.Unix.cs:
```diff
@@ -0,0 +1,57 @@
+using System.Net;
+using System.Net.Security;
+
+namespace System.Security.Cryptography.X509Certificates
```
This doesn't seems like it really belongs in this namespace (and include path), due to layering problems.  `SslPolicyErrors` is a System.Net.Security enum, so it's really a System.Net.Security-ism more than an X509-ism.

So none of the cryptography libraries could include this thing from the crypto namespace.

stephentoub replied:
I'll move it to System.Net.

bartonjs commented on src/System.Net.Security/src/System/Net/CertificateValidationPal.Unix.cs:
```diff
@@ -24,60 +24,7 @@ internal static partial class CertificateValidationPal
-                }
-            }
-            return sslPolicyErrors;
+            return CertificateValidation.BuildChainAndVerifyProperties(chain, remoteCertificate, checkCertName, hostName);
```
There shouldn't be too many callers of this, why forward like this instead of move whole-hog?

stephentoub replied:
Move it where?  It's used by both System.Net.Security and System.Net.Http.

bartonjs replied:
Oh, my bad.  This method presumably has a required name, since it's a non-private PAL method...  (the question was supposed to be "why not just update the caller(s) to call the new method and delete this old one?")

stephentoub replied:
> This method presumably has a required name, since it's a non-private PAL method

Right, there's a corresponding method with the same signature in the Windows implementation.

stephentoub commented:
Updated.  @bartonjs, can you take another pass?

bartonjs commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.ClientCertificateProvider.cs:
```diff
@@ -83,20 +109,29 @@ private int TlsClientCertCallback(IntPtr ssl, out IntPtr certHandle, out IntPtr
+                }
+                finally
+                {
+                    if (certificate != null && _clientCertificates != null) certificate.Dispose(); // only dispose cert if it's not newly created
```
I don't understand this line... you're disposing the cert if it came from a user, but letting it get finalized if it was automatic?

Since the object came from the user, they might not expect that they're giving up the lifetime, so the user-provided cert shouldn't be disposed.  (They might be passing it on the next request)

stephentoub replied:
That should be _clientCertificates == null, not _clientCertificates != null.  It should be disposed in the automatic case and not disposed in the manual case.

bartonjs commented on src/Common/src/System/Net/Security/CertificateValidation.Unix.cs:
```diff
@@ -0,0 +1,56 @@
+            }
+
+            int hostNameMatch;
+            using (SafeX509Handle certHandle = Interop.Crypto.X509Duplicate(remoteCertificate.Handle))
```
@stephentoub (Moving the name verification discussion to a non-outdated-diff :smile:)

The name verification code, which is this using, only operates on the SafeX509Handle (or, rather, the IntPtr that came in representing it originally).  So, if we wanted to always do the name management ourselves, we'd just need to extract this to a helper method which took hostName and IntPtr

stephentoub replied:
Which would still need to then get the address bytes from the IP or convert the string to ascii, and then marshal those to native code.  Less cost, but still cost.

bartonjs replied:
True, true.  I just meant it wouldn't be as expensive as the chain building operation, since that requires object hydration, filesystem enumeration, etc.  I'm fine with the approach we have here.

If it turns out that libcurl doesn't do IPAddress matching correctly, or IDNA, then maybe we'll have to do something different.

bartonjs commented on src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.ClientCertificateProvider.cs:
```diff
@@ -138,12 +168,29 @@ private static ISet<string> GetRequestCertificateAuthorities(SafeSslHandle sslHa
+                        X509Certificate2 cert = certs[i];
+                        if (cert != certificate)
+                        {
+                            cert.Dispose();
```
(Yay!)

stephentoub replied:
:smile:

bartonjs commented:
Okay, the bad dispose is my only outstanding issue.  Flip it to dispose-on-auto and it'll LGTM.

stephentoub commented:
Thanks, Jeremy.  Fixed.

bartonjs commented:
LGTM :smile:.

stephentoub commented:
@dotnet-bot test this please (failures due to GitHub briefly being down)

stephentoub commented:
(Windows failed leg is due to a test deadlock in the tasks suite. https://github.com/dotnet/corefx/pull/7107.  Since these changes are Unix only, I'm not going to re-run it.)

stephentoub commented:
@SidharthNabar, @davidsh, any concerns with my merging this?

SidharthNabar commented:
@stephentoub - Thanks for turning this around so quickly! 
Since this change is not changing the API surface in any way (yet), I am good with merging this. Note that depending on what we land on for the API surface of HttpClientHandler, we may have to modify this implementation accordingly, but it should still be a significant stepping stone towards the final version.

stephentoub commented:
Thanks, Sid.  And Jeremy.
